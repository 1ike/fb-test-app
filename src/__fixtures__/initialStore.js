const options = {
  preset: "islands#blueCircleDotIconWithCaption",
  draggable: true
};

const point1 = {
  id: 0,
  name: "p0",
  placemark: {
    geometry: {
      coordinates: [55.78, 37.63]
    },
    properties: {
      hintContent: "p0",
      balloonContent: "p0"
    },
    options
  }
};
const point2 = {
  id: 1,
  name: "p1",
  placemark: {
    geometry: {
      coordinates: [55.74, 37.62]
    },
    properties: {
      hintContent: "p1",
      balloonContent: "p1"
    },
    options
  }
};
const point3 = {
  id: 2,
  name: "p2",
  placemark: {
    geometry: {
      coordinates: [55.76, 37.7]
    },
    properties: {
      hintContent: "p2",
      balloonContent: "p2"
    },
    options
  }
};

export default {
  points: [point1, point2, point3]
};

export const reorderedPoints = {
  points: [
    JSON.parse(JSON.stringify(point2)),
    JSON.parse(JSON.stringify(point1)),
    JSON.parse(JSON.stringify(point3))
  ]
};
