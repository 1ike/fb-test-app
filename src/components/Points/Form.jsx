import React from "react";
import { Field, reduxForm } from "redux-form";

import _connect from "../../redux/connect";

const uid = (() => {
  let id = 5;
  return () => id++;
})();

const mapStateToProps = ({ center }) => ({ center });

export class Form extends React.Component {
  inputName = "inputPoint";

  createPoint = value => {
    const name = value[this.inputName];
    if (!name) {
      this.focusInput();
      return;
    }

    this.props.createPoint({
      id: uid(),
      name,
      placemark: {
        geometry: {
          coordinates: this.props.center
        },
        properties: {
          hintContent: name,
          balloonContent: name
        },
        options: {
          preset: "islands#blueCircleDotIconWithCaption",
          draggable: true
        }
      }
    });
    this.props.reset();
  };

  focusInput = () => {
    this.input.getRenderedComponent().focus();
  };

  render() {
    const title = "Добавить точку";
    return (
      <form
        id="newPointForm"
        className="new-point-form"
        onSubmit={this.props.handleSubmit(this.createPoint)}
      >
        <Field
          component="input"
          name={this.inputName}
          id={this.inputName}
          placeholder=" Введите новую точку маршрута"
          className="new-point-form__input"
          ref={input => {
            this.input = input;
          }}
          withRef
          autoFocus={true}
        />
        <button
          type="submit"
          className="new-point-form__submit-button"
          aria-label={title}
          title={title}
        >
          +
        </button>
      </form>
    );
  }
}

const reduxFormWrapper = reduxForm({ form: "createPoint" })(Form);

export default _connect(mapStateToProps)(reduxFormWrapper);
