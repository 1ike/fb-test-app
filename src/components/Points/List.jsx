import React from "react";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";

import _connect from "../../redux/connect";

const CloseButton = props => {
  const title = "Удалить точку";
  return (
    <button
      type="button"
      className="sortable-list__delete-button"
      aria-label={title}
      title={title}
      onClick={props.onClick}
    >
      <span aria-hidden="true">&times;</span>
    </button>
  );
};

// a little function to help us with reordering the result
const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

const getItemStyle = (isDragging, draggableStyle) => ({
  // change background colour if dragging
  background: isDragging ? "lightgreen" : "grey",

  // styles we need to apply on draggables
  ...draggableStyle
});

const getListStyle = isDraggingOver => ({
  background: isDraggingOver ? "lightblue" : "lightgrey"
});

const mapStateToProps = state => {
  const props = {
    points: state.points
  };
  return props;
};

class List extends React.Component {
  onDragEnd = () => result => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }

    const reoderedPoints = reorder(
      this.props.points,
      result.source.index,
      result.destination.index
    );

    this.props.reoderPoints(reoderedPoints);
  };

  deletePoint = id => () => {
    this.props.deletePoint(id);
  };

  render() {
    const points = this.props.points;
    return (
      <DragDropContext onDragEnd={this.onDragEnd()}>
        <Droppable droppableId="droppable">
          {(provided, snapshot) => (
            <div
              id="list"
              className="sortable-list"
              ref={provided.innerRef}
              style={getListStyle(snapshot.isDraggingOver)}
            >
              {points.map((point, index) => (
                <Draggable key={point.id} draggableId={point.id} index={index}>
                  {(provided, snapshot) => (
                    <div
                      id={`point_${point.id}`}
                      className="sortable-list__item"
                      ref={provided.innerRef}
                      {...provided.draggableProps}
                      {...provided.dragHandleProps}
                      style={getItemStyle(
                        snapshot.isDragging,
                        provided.draggableProps.style
                      )}
                    >
                      <span className="sortable-list__item-text">
                        {point.name}
                      </span>
                      <CloseButton onClick={this.deletePoint(point.id)} />
                    </div>
                  )}
                </Draggable>
              ))}
              {provided.placeholder}
            </div>
          )}
        </Droppable>
      </DragDropContext>
    );
  }
}

export default _connect(mapStateToProps)(List);
