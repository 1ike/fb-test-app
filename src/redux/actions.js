import { createActions } from "redux-actions";

export const {
  createPoint,
  deletePoint,
  shiftPoint,
  reoderPoints,
  shiftCenter
} = createActions(
  // {
  //   ACTION_ONE: (key, value) => ({ [key]: value })
  // },
  "CREATE_POINT",
  "DELETE_POINT",
  "SHIFT_POINT",
  "REODER_POINTS",
  "SHIFT_CENTER"
);
