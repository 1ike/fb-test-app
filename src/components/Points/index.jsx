import React from "react";

// import _connect from '../../connect';
import Form from "./Form";
import List from "./List";

export default class Points extends React.Component {
  render() {
    return (
      <div id="points">
        <Form />
        <List />
      </div>
    );
  }
}
