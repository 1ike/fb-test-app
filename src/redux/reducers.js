import { combineReducers } from "redux";
import { reducer as form } from "redux-form";
import { handleActions } from "redux-actions";

import * as actions from "./actions";

const points = handleActions(
  {
    [actions.createPoint](state, { payload: point }) {
      return [...state, point];
    },
    [actions.deletePoint](state, { payload: id }) {
      return state.filter(point => point.id !== id);
    },
    [actions.shiftPoint](
      state,
      {
        payload: { id, coordinates }
      }
    ) {
      return state.map(point => {
        if (point.id !== id) return point;
        point.placemark.geometry.coordinates = coordinates;
        return point;
      });
    },
    [actions.reoderPoints](state, { payload: points }) {
      return points;
    }
  },
  []
);

const center = handleActions(
  {
    [actions.shiftCenter](state, { payload }) {
      return payload;
    }
  },
  [55.76, 37.64]
);

export default combineReducers({
  form,
  points,
  center
});
