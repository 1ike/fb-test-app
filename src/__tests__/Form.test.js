import React from "react";
import { Provider } from "react-redux";
import { reduxForm } from "redux-form";

import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { Form } from "../components/Points/Form";

import initialStore from "../__fixtures__/initialStore";
import { createStore } from "redux";
import reducers from "../redux/reducers";

Enzyme.configure({ adapter: new Adapter() });

const inputSelector = "input#inputPoint";

const store = createStore(reducers, initialStore);
const WrappedForm = reduxForm({ form: "createPoint" })(Form);

it("Form snapshots", () => {
  const wrapper = mount(
    <Provider store={store}>
      <WrappedForm />
    </Provider>
  );
  wrapper
    .find(inputSelector)
    .simulate("change", { target: { value: "test point 1" } });
  expect(wrapper.html()).toMatchSnapshot();
});
