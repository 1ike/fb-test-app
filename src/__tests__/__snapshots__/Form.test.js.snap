// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshots 1`] = `"<form id=\\"newPointForm\\" class=\\"new-point-form\\"><input name=\\"inputPoint\\" id=\\"inputPoint\\" placeholder=\\" Введите новую точку маршрута\\" class=\\"new-point-form__input\\" value=\\"test point 1\\"><button type=\\"submit\\" class=\\"new-point-form__submit-button\\" aria-label=\\"Добавить точку\\" title=\\"Добавить точку\\">+</button></form>"`;
